pkg/yurttunnel/agent/agent.go

```diff
@@ -22,119 +22,46 @@ import (
-	}
-}
-
-// getServerAddrLoadBalancer gets the service address of the yurttunnel-server
-// if the service type is LoadBalancer
-func getServerAddrLoadBalancer(
-	svc *corev1.Service) (string, error) {
-	var tcpPort int32
-	for _, port := range svc.Spec.Ports {
-		if port.Name == constants.YurttunnelServerAgentPortName {
-			tcpPort = port.Port
-			break
-		}
-	}
-
-	for _, ingress := range svc.Status.LoadBalancer.Ingress {
-		if len(ingress.IP) != 0 {
-			return fmt.Sprintf("%s:%d", ingress.IP, tcpPort), nil
-		}
-	}
-	return "", errors.New("can't find qualified ingress")
-}
-
-// getServerAddrClusterIP gets the service address of the yurttunnel-server
-// if the service type is ClusterIP
-func getServerAddrClusterIP(
-	clientset kubernetes.Interface,
-	svc *corev1.Service) (string, error) {
-	if addr, ok := svc.Annotations[constants.YurttunnelServerExternalAddrKey]; ok {
-		return addr, nil
-	}
-
-	eps, err := clientset.CoreV1().Endpoints(constants.YurttunnelEndpointsNs).
-		Get(constants.YurttunnelEndpointsName, metav1.GetOptions{})
+	_, ips, err := certmanager.GetYurttunelServerDNSandIP(clientset)
 	if err != nil {
 		return "", err
 	}
 
-	for _, ss := range eps.Subsets {
-		if len(ss.Addresses) == 1 && len(ss.Ports) == 1 {
-			return fmt.Sprintf("%s:%d", ss.Addresses[0].IP, ss.Ports[0].Port), nil
-		}
+	if len(ips) <= 0 {
+		return "", errors.New("there is no available ip")
 	}
-	return "", errors.New("can't find qualified endpoint subsets")
-}
 
-// getServerAddrNodePort gets the service address of the yurttunnel-server
-// if the service type is NodePort
-func getServerAddrNodePort(
-	clientset kubernetes.Interface,
-	svc *corev1.Service) (string, error) {
-	// get node ip
-	labelSelector := "alibabacloud.com/is-edge-worker=false"
-	nodeLst, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{
-		LabelSelector: labelSelector,
-	})
-	if err != nil {
-		return "", err
-	}
-	if len(nodeLst.Items) == 0 {
-		return "", errors.New("there is no cloud node")
-	}
-	var (
-		nodeIP      string
-		foundNodeIP bool
-	)
-	for _, addr := range nodeLst.Items[0].Status.Addresses {
-		if addr.Type == corev1.NodeInternalIP {
-			nodeIP = addr.Address
-			foundNodeIP = true
-		}
-	}
-	if !foundNodeIP {
-		return "", errors.New("can't find node IP")
-	}
-	// get node port
-	var (
-		tcpPort      int32
-		foundTCPPort bool
-	)
+	var tcpPort int32
 	for _, port := range svc.Spec.Ports {
 		if port.Name == constants.YurttunnelServerAgentPortName {
-			tcpPort = port.NodePort
-			foundTCPPort = true
+			if svc.Spec.Type == corev1.ServiceTypeNodePort {
+				tcpPort = port.NodePort
+			} else {
+				tcpPort = port.Port
+			}
 			break
 		}
 	}
-	if !foundTCPPort {
-		return "", errors.New("tcp port not found")
+
+	if tcpPort == 0 {
+		return "", errors.New("fail to get the port number")
 	}
-	return fmt.Sprintf("%s:%d", nodeIP, tcpPort), nil
+
+	return fmt.Sprintf("%s:%d", ips[0].String(), tcpPort), nil
```

Comment:
from the return values of GetYurttunelServerDNSandIP  func, ips[0] value is 127.0.0.1, is not node ip